doc/man7/password-handling.pod

```diff
@@ -0,0 +1,89 @@
+=pod
+
+=head1 NAME
+
+password handling - How diverse parts of OpenSSL treat pass phrases
+
+=head1 DESCRIPTION
+
+=head2 The general case
+
+In a modern world with all sorts of character encodings, the treatment of pass
+phrases has become increasingly complex.
+
+The OpenSSL library doesn't treat pass phrases in any special way as a general
+rule, and trusts the application or user to choose a suitable character set
+and stick to that throughout the life of affected objects.
+This means that for an object that was encrypted using a pass phrase encoded in
+ISO-8859-1, that object needs to be decrypted using a pass phrase encoded in
+ISO-8859-1.
+Anything else will have the decryption fail.
+
+=head2 PKCS#12
+
+PKCS#12 is a bit different regarding pass phrase encoding.
+The standard stipulates that the pass phrase shall be encoded as a BMPString,
+i.e. UTF-16.
+
+OpenSSL tries to adapt to this requirements in one of the following manners:
+
+=over 4
+
+=item 1.
+
+Tries to read the pass phrase as UTF-8 and re-encode that to UTF-16, or failing
+that...
```

Comment:
Using a  continuation marker across numbered items...  (And continuing it with a capital letter) feels wrong to me. How about

> , or failing that, proceeds with step 2.

Reply:
Ah, yes, I did thinking that the way I wrote it was...  a bit clumsy.  Good idea

Reply:
There is ambiguity in wording "read the pass phrase". It might give an impression that input is actually read, from prompt or file, in subroutine in question, while it's read elsewhere. Here the string is only **interpreted**. Why is it significant? Because elsewhere plays role too, and important one... But back to PKCS#12 module. It attempts to interpret input as UTF-8, and falls back to 1.0.x way if it's not. Relevant question is what would it take to trigger the fallback for an iso8859-X string? *Any* transition from last quadrant of the the table to first half, a.k.a. ASCII, triggers the fallback, e.g. `ïv` in iso8859-1 `naïve`. *Any* pair of characters from last quadrant triggers the fallback, e.g. any pair in iso8859-7 `σύνθημα γνώρισμα`. *Any* character from last quadrant at the end of the input string triggers the fallback, e.g. `ö` in iso8859-1 `Malmö`. In practical terms this means that if you have PKCS#12 file protected with passphrase originally entered under non-UTF-8 locale, all it takes is to start terminal prompt with original locale and off it goes. Well, there still is possibility that fallback won't be triggered. Because there are combinations of non-UTF-8 characters that would make PKCS#12 code recognize string as UTF-8 and *not* resort for 1.0.2 way. Which are those? Basically pairs of characters from last quadrant and one before last, possibly among characters from first half. For example modified `naï¼ve` would be mistreated. (Note that there is no way to mess up `σύνθημα γνώρισμα` with additional characters, or `Malmö` as long as it ends with `ö`.) But even if `naï¼ve` would be mistaken for UTF-8 input, would openssl pkcs12 fail? No (at at least it shouldn't, as it was part of my original submission). Because misinterpreted string won't decrypt the object and second attempt will be taken, now straight 1.0.2 way, without relying on above mentioned fallback mechanism. In other words there is kind of double-fallback. Bottom line is that Unix users are in pretty good shape recovering PKCS#12 files from 90-s protected with iso8859-X passphrases. But they have to recall original locale and know how to open terminal prompt with it. This is the "elsewhere" mentioned in the beginning. Well, of course this is just one corner case, let's call it "one from 90-s". Couple of others will be discussed separately...

Just in case for reference. Last quadrants of iso8859-X tables customarily house "printables", characters that are found in the "middle" of keyboard. However, in some tables there are "printables" even in one before last quadrant. So that one can actually find problematic yet "pronounceable" combinations, e.g. `īŋ` in iso8859-4, that would evade the fallback.

Reply:
Another corner case is from this century, when vast majority of Unix users are more than likely to use UTF-8 [input] locales. What are problems there? As correctly mentioned 1.0.x takes whichever input and prepends each character with 0. So that UTF-8 `naïve` would be [**incorrectly**] converted to 006e006100c300af00760065 and used as password to protect PKCS#12 object. 1.1.x on the other hand will convert it [**correctly**] to 006e006100ef00760065. But recall that it's actually double-fallback! 1.1.x openssl pkcs12 is supposed to try both strings, in correct-incorrect order. So that Unix users can recover malformed PKCS#12 files from this century. Note that there is kind of overlap with the case from 90-s. Imagine 1.0.x[!] user with file protected with iso8859-X passphrase switching to UTF-8 locale. It actually wouldn't work, and user would have to resort to original iso8859-X input locale. This means that users with files "from 90-s" has to know this since forever. And the trick keeps working with 1.1.x, which gives consistent view. Files "from 90-s" -> just stick to original input locale.

Reply:
So far remarks were referring explicitly to Unix users. That's because another corner case of its own is ... Windows. Even though there are iso8859-X code pages (I think there are), by default Windows use custom ones. It's just safer to assume that PKCS#12 files protected with non-ASCII passphrases by pre-1.1.x on Windows are **not interoperable** with **anything**. It's just the world of its own. But what could 1.1.x do? There is no UTF-8 *input* locale... On Windows [this](https://github.com/openssl/openssl/blob/master/CHANGES#L675) happens. There is no magic, user just has to make active choice. Not setting environment variable allows you to work with pre-1.1.x files, and setting one allows you to interoperate with other software including 1.1.x OpenSSL on Unix. In the nutshell most appropriate thing user can do is to convert pre-1.1.x file to 1.1.x and stick to environment variable for the eternity...